File: 151051/Maliuk/ipr/src/main/java/ru/bsuir/services/CreatorService.java
package ru.bsuir.services;

import lombok.AllArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.cache.annotation.EnableCaching;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import ru.bsuir.dto.request.CreatorRequestTo;
import ru.bsuir.dto.response.CreatorResponseTo;
import ru.bsuir.entity.Creator;
import ru.bsuir.irepositories.ICreatorRepository;
import ru.bsuir.mapper.CreatorMapper;

import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
@EnableCaching
public class CreatorService {

    private final ICreatorRepository creatorRepository;
    private final CreatorMapper creatorMapper;

    @CacheEvict(value = "creators", allEntries = true)
    public CreatorResponseTo createCreator(CreatorRequestTo creatorRequest) {

        Creator creator = creatorMapper.toEntity(creatorRequest);
        creatorRepository.save(creator);
        return creatorMapper.toDTO(creator);
    }

    @Cacheable(value = "creators", key = "#id")
    public CreatorResponseTo getCreatorById(Long id) {
        Optional<Creator> creatorOpt = creatorRepository.findById(id);
        return creatorOpt.map(creatorMapper::toDTO)
                .orElseThrow(() -> new RuntimeException("Creator not found"));

    }

    @Cacheable(value = "creatorsList")
    public List<CreatorResponseTo> getAllCreator(){
        return  creatorRepository.findAll().stream()
                .map(creatorMapper::toDTO)
                .toList();
    }


    @CacheEvict(value = {"creators", "creatorsList"}, key = "#id", allEntries = true)
    public CreatorResponseTo updateCreator(Long id, CreatorRequestTo creatorRequest) {
        Creator creator = creatorRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Creator not found"));

        if (creatorRequest.getLogin().length() < 2) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Login must be at least 2 characters long");
        }
        creator.setLogin(creatorRequest.getLogin());
        creator.setPassword(creatorRequest.getPassword());
        creator.setFirstname(creatorRequest.getFirstname());
        creator.setLastname(creatorRequest.getLastname());

        creatorRepository.save(creator);
        return creatorMapper.toDTO(creator);
    }

    @CacheEvict(value = {"creators", "creatorsList"}, key = "#id", allEntries = true)
    public void deleteCreator(Long id) {

        if(!creatorRepository.existsById(id)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Creator not found");
        }
        creatorRepository.deleteById(id);
    }
    public boolean existById(Long id) { return creatorRepository.existsById(id);}
}

File: 251002/Nikonorova/lab1_rest/Services/CreatorService.cs
using distcomp.DTOs;
using distcomp.Models;
using distcomp.Repositories;
using Microsoft.AspNetCore.Http.HttpResults;
using System.Collections.Generic;

namespace distcomp.Services
{
    public class CreatorService
    {
        private readonly CreatorRepository _creatorRepository;

        public CreatorService(CreatorRepository creatorRepository)
        {
            _creatorRepository = creatorRepository;
        }

       
        public CreatorResponseTo CreateCreator(CreatorRequestTo creatorRequest)
        {
            var creator = new Creator
            {
                Login = creatorRequest.Login,
                Password = creatorRequest.Password,
                Firstname = creatorRequest.Firstname,
                Lastname = creatorRequest.Lastname
            };

            var savedCreator = _creatorRepository.Save(creator);
            return MapToCreatorResponse(savedCreator);
        }

        
        public CreatorResponseTo GetCreatorById(long id)
        {
            var creator = _creatorRepository.FindById(id);
            return creator != null ? MapToCreatorResponse(creator) : null;
        }

        
        public List<CreatorResponseTo> GetAllCreators()
        {
            var creators = _creatorRepository.FindAll();
            var creatorResponses = new List<CreatorResponseTo>();


            foreach (var creator in creators)
            {
                creatorResponses.Add(MapToCreatorResponse(creator));
            }
            return creatorResponses;
        }


        
        public CreatorResponseTo UpdateCreator(long id, CreatorRequestTo creatorRequest)
        {
            var creator = _creatorRepository.FindById(id); //checking that user exists
            if (creator == null)
                return null;

            creator.Login = creatorRequest.Login;
            creator.Password = creatorRequest.Password;
            creator.Firstname = creatorRequest.Firstname;
            creator.Lastname = creatorRequest.Lastname;

            var updatedCreator = _creatorRepository.Update(creator);
            return MapToCreatorResponse(updatedCreator);
        }

       
        public bool DeleteCreator(long id)
        {
            return _creatorRepository.DeleteById(id);
        }

        // transforming user entity into DTO
        private CreatorResponseTo MapToCreatorResponse(Creator creator)
        {
            return new CreatorResponseTo
            {
                Id = creator.Id,
                Login = creator.Login,
                Firstname = creator.Firstname,
                Lastname = creator.Lastname
            };
        }
    }
}
